[terra-slider] New Terra slider component creation

nikhitasharma opened the pull request:
<!-- 
PLEASE COMPLETE THESE STEPS BEFORE PUBLISHING THE PULL REQUEST:

*Before publishing*
1. Update the title as: [component] changes to component (ex. [terra-button] Added terra-button accessibility guide).
2. Fill out the fields below.
3. Assign yourself to the PR.
4. Add the appropriate package name labels.
5. Add your name to the CONTRIBUTORS.md file. Adding your name to the CONTRIBUTORS.md file signifies agreement to all rights and reservations provided by the License.
-->

### Summary
<!--- Summarize and explain the reasoning behind these code changes. What are the changes, and why are they necessary? -->

**What was changed:**
Adding new Terra Slider component along with documentation



### Testing
<!-- Demonstrate that these changes are stable. How have these changes been verified? -->

This change was tested using:

- [x] WDIO
- [x] Jest
- [x] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:
<!-- Please include the appropriate "Required" & "Ready" labels (ex. "UX Review Required" and "UX Review Ready").  -->
- [x] UX review
- [x] Accessibility review
- [x] Functional review

### Additional Details
<!-- List anything else that is relevant to this issue. Additional information will help us better understand your changes and speed up the review process. -->

**This PR resolves:**

UXPLATFORM-9555
UXPLATFORM-9600

---

Thank you for contributing to Terra.
@cerner/terra

sycombs commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+      <label className={cx('label')} id="sliderLabel">
+        {labelName}
+      </label>
+      <span className={cx('min-label')}>
+        {minLabel}
+      </span>
+      <input className={cx('input-range')} type="range" ref={sliderRef} aria-labelledby="sliderLabel" aria-valuemin={minimumValue} aria-valuemax={maximumValue} aria-valuenow={defaultValue} defaultValue={defaultValue} disabled={isDisabled} min={minimumValue} max={maximumValue} onChange={handleValueChange} />
```
This is a lot of attributes on one line, can we break them up so it's more readable?

```suggestion
      <input
        className={cx('input-range')}
        type="range"
        ref={sliderRef}
        aria-labelledby="sliderLabel"
        aria-valuemin={minimumValue}
        aria-valuemax={maximumValue}
        aria-valuenow={defaultValue}
        defaultValue={defaultValue}
        disabled={isDisabled}
        min={minimumValue}
        max={maximumValue}
        onChange={handleValueChange}
      />
```

cm9361 replied:
Should we be leveraging our Terra Arrange component for the min label, slider, max label layout?

sycombs commented on packages/terra-slider/translations/en-US.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "Terra.slider.Slider": "Slider"
```
Is there a JIRA logged to add the translations for other locales?

nikhitasharma replied:
There is no Jira logged for this currently, maybe because it was not a requirement for the passthrough. I agree that one should be logged and will follow-up on this and post here.

nikhitasharma replied:
Thank you @chrismichalewicz!

cm9361 replied:
Where would we be using this translation?

sycombs commented on packages/terra-slider/tests/jest/Slider.test.jsx:
```diff
@@ -0,0 +1,25 @@
+    const wrapper = shallow(
+      <Slider isDisabled minimumValue={0} maximumValue={100} labelName="testLabel" defaultValue={50} />,
+    );
+    expect(wrapper).toMatchSnapshot();
```
Can we add an assertion to check the value of the `disabled` attribute on the `input` element?

sycombs commented on packages/terra-slider/tests/jest/Slider.test.jsx:
```diff
@@ -0,0 +1,25 @@
+    const wrapper = shallow(
+      <Slider minimumValue={0} maximumValue={100} labelName="testLabel" minimumLabel="testMinimumLabel" maximumLabel="testMaximumLabel" defaultValue={50} />,
+    );
+    expect(wrapper).toMatchSnapshot();
```
Similar to the above comment, can we add an assertion to check the values of the minimum and maximum labels?

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+  /**
+  * Default or initial value for the slider.
+  */
+  defaultValue: PropTypes.number.isRequired,
```
This property should probably be called, value.  We want the consumer to be able to specify the value of the component not the default value.

nikhitasharma replied:
Will update this.

nikhitasharma replied:
Updated.

cm9361 commented:
We should probably create a test and example where the consumer is changing the slider value via the prop.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    width: 20%;
+  }
+
+  .min-label {
+    color: #6f7477;
+    font-size: var(--terra-slider-range-label-font-size, 12px);
+    font-weight: 400;
+    margin-right: 7px;
+    text-align: left;
+    white-space: initial;
+  }
+
+  .max-label {
+    color: #6f7477;
+    font-size: var(--terra-slider-range-label-font-size, 12px);
+    font-weight: 400;
+    margin-left: 7px;
+    text-align: left;
+    white-space: initial;
+  }
```
You have the same values for both classes.  You should just create one class for both the min and max labels.

nikhitasharma replied:
We do not have same values for these classes, the margin needs to be on the right for min label and on the left for max label.

cm9361 replied:
Then you should pull out that one style instead of duplicating all of them.

nikhitasharma replied:
Updated.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    width: 20%;
+  }
+
+  .min-label {
+    color: #6f7477;
+    font-size: var(--terra-slider-range-label-font-size, 12px);
+    font-weight: 400;
+    margin-right: 7px;
+    text-align: left;
+    white-space: initial;
+  }
+
+  .max-label {
+    color: #6f7477;
+    font-size: var(--terra-slider-range-label-font-size, 12px);
+    font-weight: 400;
+    margin-left: 7px;
+    text-align: left;
+    white-space: initial;
+  }
```
You have the same values for both classes.  You should just create one class for both the min and max labels.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+
+  .label {
+    color: var(--terra-slider-label-color, #1c1f21);
+    font-weight: 700;
+    margin-right: 7px;
```
You should have theme variables from the font weight and margin

nikhitasharma replied:
These values are same in all themes, I don't see a need to have theme variables for these.

cm9361 replied:
These are not the only themes.  We have the Redwood theme coming up as well.  There is also the cerner-clinical-theme.  We cannot limit to just designing for lowlight and fusion.

nikhitasharma replied:
Updated.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  .slider {
+    align-items: center;
+    display: inline-flex;
+    min-height: 44px;
```
You should have a theme variable for this value.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  .min-label {
+    color: #6f7477;
```
This makes me think the lowlight theme is not being handled properly.  There should be a theme variable as I would not expect the font color to be the same for lowlight and light themes.  It should also be configurable for other themes.

nikhitasharma replied:
These are infact given same values by UX. I agree with you though. I will double check with them on this and if we need to change them to be different.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  .min-label {
+    color: #6f7477;
+    font-size: var(--terra-slider-range-label-font-size, 12px);
+    font-weight: 400;
+    margin-right: 7px;
```
You should have theme variables for these styles.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  /* For IE/Edge */
+  .input-range::-ms-track {
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background: linear-gradient(to right, var(--terra-slider-track-selected-range, #007cc3) 0%, var(--terra-slider-track-selected-range, #007cc3) var(--terra-slider-progress-status, 50%), var(--terra-slider-track-unselected-range, #d3d4d5) var(--terra-slider-progress-status, 50%), var(--terra-slider-track-unselected-range, #d3d4d5) 100%);
```
I would expect this to be one single theme variable.

cm9361 replied:
There just seems to be too many variables here.  A consumer would just set the linear gradient.  They don't need all of these variables to do that.  It should be done with one variable.

nikhitasharma replied:
Discussed this offline and explained the progress status updates for slider css. Concluded that this cannot be a single theme variable because of dynamic updates and will keep this as is. Any further css optimizations will be investigated in UXPLATFORM-9778.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  .input-range::-ms-track {
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background: linear-gradient(to right, var(--terra-slider-track-selected-range, #007cc3) 0%, var(--terra-slider-track-selected-range, #007cc3) var(--terra-slider-progress-status, 50%), var(--terra-slider-track-unselected-range, #d3d4d5) var(--terra-slider-progress-status, 50%), var(--terra-slider-track-unselected-range, #d3d4d5) 100%);
+    border-radius: 3px;
+    height: 6px;
+    width: 6px;
```
You should have theme variables for these values.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
You have a lot of styles with no theme variables.  You need to identify the set of variables necessary and ensure that the styles can be themed.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background: var(--terra-slider-thumb-background, #007cc3);
+    border-color: var(--terra-slider-thumb-border-color, #0065a3);
+    border-radius: 50%;
```
Did the border radius of 50% come from a UX callout?  Usually the value is in rem units.

nikhitasharma replied:
Updated to rems.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background: var(--terra-slider-thumb-background, #007cc3);
+    border-color: var(--terra-slider-thumb-border-color, #0065a3);
+    border-radius: 50%;
+    border-style: solid;
+    border-width: 1px;
+    box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.25);
+    cursor: pointer;
+    height: 24px;
+    margin-top: -8px;
+    width: 24px;
```
These should be themeable.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  .input-range:focus::-webkit-slider-thumb {
+    border-color: var(--terra-slider-thumb-focus-border-color, #007cc3);
+    border-width: 1px;
```
Why not just use the border style?
border: 1px solid var(...)

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  .input-range:hover::-webkit-slider-thumb {
+    background: var(--terra-slider-thumb-hover-background, #0065a3);
+    border-color: var(--terra-slider-thumb-hover-border-color, #0065a3);
```
In your logic the background color and border color are the same.  Do you just not need a border?

nikhitasharma replied:
these were same for hover for some reason in the call-outs. I will double check with UX on this.

cm9361 replied:
I think some of the newer PRs used "background".  You definitely don't need both styles though.  We don't need to consult with UX on that.  Using background would allow other themes to use a gradient color, if desired.

eawww replied:
@cm9361 this is probably to ensure the border color is overridden for the hover style. There's several ways this could be accomplished but this seems alright.

cm9361 replied:
@eawww You only need one of the styles.  The background style can be set to a color.  We don't want to have two styles that can conflict each other with new code.

nikhitasharma replied:
I have removed the border color wherever they are same as the background color. I checked that there are no visual problems or thumb size changes after removing this.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  .input-range:hover::-webkit-slider-runnable-track {
+    background: linear-gradient(to right, var(--terra-slider-track-selected-range-hover, #0065a3) 0%, var(--terra-slider-track-selected-range-hover, #0065a3) var(--terra-slider-progress-status, 50%), var(--terra-slider-track-unselected-range-hover, #bcbfc0) var(--terra-slider-progress-status, 50%), var(--terra-slider-track-unselected-range-hover, #bcbfc0) 100%);
```
I would expect this to be a single theme variable.  Same comment for the other gradient backgrounds.  The current API is too complicated.

cm9361 replied:
It doesn't appear that this was updated.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    background: var(--terra-slider-thumb-hover-background, #0065a3);
+  }
+
+  .input-range:active::-ms-thumb {
+    background: var(--terra-slider-thumb-active-background, #007cc3);
+  }
```
The colors for focus and active are the same.  You should be able to combine those styles in your stylesheet.

nikhitasharma replied:
Updated these colors. They are different now.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+  /**
+  * Minimum value label to display on the left side of the slider.
+  */
+  minimumLabel: PropTypes.string,
+
+  /**
+  * Maximum value label to display on the right side of the slider.
+  */
+  maximumLabel: PropTypes.string,
```
Do we need a minimum label and maximum label?  Isn't the requirement to show the actual values?  We already have props for the minimum and maximum value.

nikhitasharma replied:
It was agreed that the minimum labels and maximum labels would be optional ways to display customized or more readable labels. These are needed to achieve that.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+  }, []);
+
+  function handleValueChange(event) {
+    event.currentTarget.style.setProperty('--terra-slider-progress-status', `${event.currentTarget.value}%`);
+    event.currentTarget.setAttribute('aria-valuenow', `${event.currentTarget.value}`);
```
This is not the proper way to update the attributes in React.

nikhitasharma replied:
Do you have a suggested way of updating the slider progress here.

cm9361 replied:
Yes, use state variables.

nikhitasharma replied:
Updated.

cm9361 commented:
You have a blue border around your slider when it has focus.  The UX callouts do not specify this behavior.  It seems like the border should not exists and the outline just around the slider thumb.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+      <span className={cx('min-label')}>
+        {minLabel}
+      </span>
+      <input className={cx('input-range')} type="range" ref={sliderRef} aria-labelledby="sliderLabel" aria-valuemin={minimumValue} aria-valuemax={maximumValue} aria-valuenow={defaultValue} defaultValue={defaultValue} disabled={isDisabled} min={minimumValue} max={maximumValue} onChange={handleValueChange} />
```
You shouldn't need aria-valuemin and aria-valuemax since you are using the native input control.  We shouldn't need aria-valuenow once we update the implementation to allow consumers to pass the value.

nikhitasharma replied:
It was confirmed in a UX/ accessibility meeting that these values need to be present for the slider. I will double check if these are required.

cm9361 replied:
We should be able to read the ARIA and APG documentation for this.  I would expect to get this for free for a semantic HTML element.

cm9361 replied:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/slider_role

HTML's <input type="range"> implicitly has the role of slider. Do not use aria-valuemax or aria-valuemin attributes on <input type="range"> elements; use min and max instead. Otherwise, any global aria-* attributes and any other aria-* attributes applicable to the slider role.

nikhitasharma replied:
@eawww @chrismichalewicz  Can you confirm that these aria props can be removed.

nikhitasharma replied:
I will do a quick change and see if it effects the screenreader behavior after removing them and post here.

cm9361 replied:
The documentation states that they should not be there.  We should not need additional confirmation.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    margin-right: 7px;
+    text-align: center;
+    white-space: initial;
+    width: 20%;
```
Why is the width 20%.  Wouldn't a label just take up the space necessary?  This seems like an i18n issue.

nikhitasharma replied:
Mentioned in callouts. The label will wrap to the next line in case of long label text. Like here:
https://github.com/cerner/terra-framework/pull/1822/files#diff-20e91930c1eda6dd71cdd1e5ab6f0cf0631a00ffe4ea4c2dc02f2a9c06eaaf0d

![image](https://github.com/cerner/terra-framework/assets/19804033/4ad6aea9-905a-473d-98bb-57f42db84cb8)

cm9361 replied:
You don't have the percentages set for the other components though.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
A lot of your style values are in pixels.  However, we usually are supposed to leverage rem units for accessibility.

nikhitasharma replied:
I will reach out to UX and see if these can be updated in the callouts to use rems.

nikhitasharma replied:
Updated to rems.

nikhitasharma commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+    width: 70%;
+  }
+  
+  .input-range:disabled {
```
This particular disabled css can be nested under input-range and I will nest them. But the rest of the browser specific css need to be separate as Firefox gives compatibility issues  when seeing styles for chrome (webkit styles). Thats why those are written separately.

cm9361 replied:
I am not clear on how that would impact the nesting approach.  It would give the exact same CSS.

nikhitasharma replied:
Came across this when firefox styling were not working when nested with chrome webkit stylings: https://stackoverflow.com/questions/48648530/moz-styling-doesnt-work-with-additional-webkit-selector

nikhitasharma commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+
+  return (
+    <div className={sliderClassNames}>
+      <label className={cx('label')} id="sliderLabel">
```
Sorry, this was meant to be removed in the finalized code.

eawww replied:
Agreed. This is causing all of the sliders on the page to read with the value of the first label.

nikhitasharma replied:
Removed.

chrismichalewicz commented:
@nikhitasharma  I performed functional testing and a11y testing and have the below comments. Items bolded I think might be issues. Then I have some general questions/comments at the bottom. 

Functionally:
- Was able to move the slider thumb as expected with mouse and keyboard arrow keys, page up/down 
- Was unable to manipulate the disabled slider.
- **In the disabled slider example when I hovered over the thumb, my mouse pointer changed to the pointing hand cursor indicating I could click on it. Since it's disabled I wouldn't expect the mouse cursor to change. @eawww please confirm.** 

Accessibility:
Keyboard:
- Was able to navigate appropriately via keyboard.
Screen Reader: (tested with JAWS in Chrome)
- Communicated "default label 0 50 100" @eawww should we indicate for 0 and 100 those are the min/max values? 
- **If I hold down the arrow keys the slides increments continuously but when I lift the key it repeates the position I am on multiple times. This seems like an issue.** 
- **In the custom slider example it communicates "0% 20 100%" I would think we'd also want the thumb value to be a percentage in this case.**  

General Question:

- Is the custom slider example to illustrate the consumer can define the slider to be something other than just integers in this example percentages? Is there other functionality we want to call out. 
- If possible I'd update the Labels to be something not generic, currently it reads as "Default Label"
- We probably need another Jira for "Implementation Guide creation" to beef up the Doc site. I can get that logged.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,212 @@
+
+  .input-range:disabled::-moz-range-thumb {
+    -moz-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background: #007cc3;
```
There should be a theme variable for this color.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+  /**
+  * Slider label value.
+  */
+  labelName: PropTypes.string.isRequired,
```
Should this prop be named something like labelText?  Don't we have that same name for some other Terra components?

nikhitasharma replied:
I am sure I saw labelName somewhere and decided to use that. But I did see labelText as well in terra-core and other repos. So, I will update this.

nikhitasharma replied:
Updated to labelText.

eawww commented on packages/terra-framework-docs/src/terra-dev-site/doc/slider/Examples.2/CustomSlider.jsx:
```diff
@@ -0,0 +1,8 @@
+import Slider from 'terra-slider';
+
+const CustomSlider = () => (
+  <Slider minimumValue={0} maximumValue={100} labelName="Field Label" minimumLabel="0%" maximumLabel="100%" defaultValue={20} />
```
We should avoid using dummy values like "Label" in our examples. Something like "Zoom", "Size", "Intensity" or something more natural like that would be better.

nikhitasharma replied:
Updated.

eawww commented on packages/terra-framework-docs/src/terra-dev-site/doc/slider/Examples.2/DefaultSlider.jsx:
```diff
@@ -0,0 +1,8 @@
+import Slider from 'terra-slider';
+
+const DefaultSlider = () => (
+  <Slider minimumValue={0} maximumValue={100} labelName="Label" defaultValue={50} />
```
At least one of these examples should show how to use the callback and value props to manage state.

nikhitasharma replied:
Added an example.

eawww reviewed (commented):
- When disabled, the slider shouldn't get the pointer cursor. It should have the default cursor.
- There should be an example that doesn't display the min/max values.
- At least one example should show how the component is used with the `onChange` and `value` props to manage application state.

chrismichalewicz commented:
> * When disabled, the slider shouldn't get the pointer cursor. It should have the default cursor.
> * There should be an example that doesn't display the min/max values.
> * At least one example should show how the component is used with the `onChange` and `value` props to manage application state.

@eawww Maybe I misinterpretted the designs that Dee put together for the min/max values I was under the impression we should always be displaying those. If that is the case would we still need an example without them?   I could see an example with non-numeric values maybe XS->XL

nikhitasharma commented on packages/terra-framework-docs/src/terra-dev-site/doc/slider/Examples.2/CustomSlider.jsx:
```diff
@@ -0,0 +1,8 @@
+import Slider from 'terra-slider';
+
+const CustomSlider = () => (
+  <Slider minimumValue={0} maximumValue={100} labelName="Field Label" minimumLabel="0%" maximumLabel="100%" defaultValue={20} />
```
Updated.

nikhitasharma commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,110 @@
+  ]);
+
+  const minLabel = minimumLabel || minimumValue;
+  const maxLabel = maximumLabel || maximumValue;
```
@eawww  Yes, they are controlled separately. The underlying values are still numeric. 
@cm9361  We need this to have the label customizations.

nikhitasharma commented:
> @nikhitasharma I performed functional testing and a11y testing and have the below comments. Items bolded I think might be issues. Then I have some general questions/comments at the bottom.
> 
> Functionally:
> 
> * Was able to move the slider thumb as expected with mouse and keyboard arrow keys, page up/down
> * Was unable to manipulate the disabled slider.
> * **In the disabled slider example when I hovered over the thumb, my mouse pointer changed to the pointing hand cursor indicating I could click on it. Since it's disabled I wouldn't expect the mouse cursor to change. @eawww please confirm.**
> 
> Accessibility: Keyboard:
> 
> * Was able to navigate appropriately via keyboard.
>   Screen Reader: (tested with JAWS in Chrome)
> * Communicated "default label 0 50 100" @eawww should we indicate for 0 and 100 those are the min/max values?
> * **If I hold down the arrow keys the slides increments continuously but when I lift the key it repeates the position I am on multiple times. This seems like an issue.**
> * **In the custom slider example it communicates "0% 20 100%" I would think we'd also want the thumb value to be a percentage in this case.**
> 
> General Question:
> 
> * Is the custom slider example to illustrate the consumer can define the slider to be something other than just integers in this example percentages? Is there other functionality we want to call out.
> * If possible I'd update the Labels to be something not generic, currently it reads as "Default Label"
> * We probably need another Jira for "Implementation Guide creation" to beef up the Doc site. I can get that logged.

- I have updated the disabled slider to not show the cursor pointer when hovered. 
- The repetitions should not happen after removing the aria attributes. Please have a look. 
- The percentage was added only to show the customization of min-max labels as that example was discussed in the meeting with UX. The underlying actual values are still integers. I am not sure there is a simple way of announcing the slider values as percentages based on custom input labels. 
- Updated labels to better values instead of "Default label"

nikhitasharma commented:
> You have a blue border around your slider when it has focus. The UX callouts do not specify this behavior. It seems like the border should not exists and the outline just around the slider thumb.

There is a consistent border color for all states, including focus. I have removed styling wherever overriding is unnecessary and if background color and border colors are the same.

eawww commented:
> In the disabled slider example when I hovered over the thumb, my mouse pointer changed to the pointing hand cursor indicating I could click on it. Since it's disabled I wouldn't expect the mouse cursor to change. Please confirm.

This is correct. Disabled slider should have the default cursor so people don't get the impression it's interactable.

nikhitasharma commented:
> We should probably create a test and example where the consumer is changing the slider value via the prop.

Added.

eawww commented:
I chatted with @scottwilmarth about some ideas to improve the screen reader experience and here's what we came up with:

What we should do is apply aria-hidden to all the visible labels including the min and max because it's all just kind of noise since we have the aria-label on the slider itself.

Then below the slider have visually hidden text that says something like

"Adjust slider to select a value between [min] ([min value]) and [max] ([max value])"

for example "Adjust slider with arrow keys to select a value between cold (0) and hot (100)" then have an aria-describedby on the slider with the id of that text.

If there are no custom labels, it can just use the min and max numbers: "Adjust slider with arrow keys to select a value between 0 and 20"

sycombs commented on packages/terra-slider/README.md:
```diff
@@ -0,0 +1,38 @@
+[![NPM version](https://badgen.net/npm/v/terra-slider)](https://www.npmjs.com/package/terra-slider)
+[![Build Status](https://badgen.net/travis/cerner/terra-framework)](https://travis-ci.com/cerner/terra-framework)
+
+Terra Slider component provides users a way to select a value from a given range by sliding a thumb along a bar to track the value.
```
small nit on wording:

```suggestion
Terra Slider provides a way to select a value from a given range by sliding a thumb along a bar to track the value.
```

nikhitasharma replied:
Updated.

sycombs commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,124 @@
+  /**
+  * Slider label value.
+  */
+  labelText: PropTypes.string.isRequired,
```
Should this just be called `label` since "text" would be implied?

nikhitasharma replied:
Updated.

sycombs commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,124 @@
+  onChange: null,
+};
+
+class Slider extends React.Component {
```
Is there a reason this is a class component? I was under the impression that any new Terra components would be functional components.

cm9361 replied:
+1.  This should definitely be a functional component.  I believe this was a comment from the previous review as well.

nikhitasharma replied:
Updated

sycombs commented on packages/terra-framework-docs/src/terra-dev-site/doc/slider/Examples.2/SliderWithOnChangeCallback.jsx:
```diff
@@ -0,0 +1,26 @@
+  const message = value ? `Updated Slider value is: ${value}` : 'Change the value of Slider to trigger callback';
+
+  return (
+    <div>
+      <Slider
+        minimumValue={0}
+        maximumValue={100}
+        labelText="Zoom"
+        value={20}
+        onChange={handleChange}
+      />
+      <p>{message}</p>
+    </div>
```
nit: Can we use a React Fragment here instead of a div?

```suggestion
    <>
      <Slider
        minimumValue={0}
        maximumValue={100}
        labelText="Zoom"
        value={20}
        onChange={handleChange}
      />
      <p>{message}</p>
    </>
```

nikhitasharma replied:
Updated.

nikhitasharma commented:
> I chatted with @scottwilmarth about some ideas to improve the screen reader experience and here's what we came up with:
> 
> What we should do is apply aria-hidden to all the visible labels including the min and max because it's all just kind of noise since we have the aria-label on the slider itself.
> 
> Then below the slider have visually hidden text that says something like
> 
> "Adjust slider to select a value between [min] ([min value]) and [max] ([max value])"
> 
> for example "Adjust slider with arrow keys to select a value between cold (0) and hot (100)" then have an aria-describedby on the slider with the id of that text.
> 
> If there are no custom labels, it can just use the min and max numbers: "Adjust slider with arrow keys to select a value between 0 and 20"

@eawww  I have added these, please have a look.

nikhitasharma commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,124 @@
+
+const propTypes = {
+  /**
+  * Slider label value.
```
Updated.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+
+const defaultProps = {
+  isDisabled: false,
+  onChange: null,
```
There is no need to set the default prop to null.  An undefined value would be falsey as well.

nikhitasharma replied:
Updated.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+    const minLabel = minimumLabel || minimumValue;
+    const maxLabel = maximumLabel || maximumValue;
+    const VisuallyHiddenTextValue = (minimumLabel || maximumLabel)
+      ? `Adjust slider to select a value between ${minimumLabel} (${minimumValue}) and ${maximumLabel} (${maximumValue})`
+      : `Adjust slider to select a value between ${minimumValue} and ${maximumValue}`;
```
You have user-visible text but no internationalization.

nikhitasharma replied:
UXPLATFORM-9721 logged to take care of internationalization.

sycombs replied:
The text should still be internationalized and these strings should be added in the en.json and en-US.json files. That Jira is for updating the translations files for the other locales.

nikhitasharma replied:
My bad, I thought these updates would take place as part of that Jira. I will get this updated soon.

nikhitasharma replied:
Added.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+
+    return (
+      /* eslint-disable-next-line react/forbid-dom-props */
+      <div style={{ '--terra-slider-progress-status': `${this.state.value}%` }} className={sliderClassNames}>
```
You should not need a CSS variable here.  You would set the appropriate style directly.

nikhitasharma replied:
Discussed offline and this is a required part of slider progress logic.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+
+Slider.propTypes = propTypes;
+Slider.defaultProps = defaultProps;
+Slider.contextType = ThemeContext;
```
You should be leveraging useContext instead of contextType to get the ThemeContext once you create your functional component.

nikhitasharma replied:
Still working on converting this to a functional component. Will get this updated soon.

nikhitasharma replied:
Updated.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+        <span className={cx('slider-label', 'slider-max-label')} aria-hidden="true">
+          {maxLabel}
+        </span>
+        <VisuallyHiddenText id={descriptionId} text={VisuallyHiddenTextValue} />
```
We should not need an id attribute for the VisuallyHiddenText element.

nikhitasharma replied:
Discussed this offline, the id attribute is required when using aria-describedby.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+
+    const minLabel = minimumLabel || minimumValue;
+    const maxLabel = maximumLabel || maximumValue;
+    const VisuallyHiddenTextValue = (minimumLabel || maximumLabel)
```
A capital letter for the beginning of a variable name is reserved for React components.

nikhitasharma replied:
Updated this.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+
+  .input-range:focus::-moz-range-thumb {
+    outline: var(--terra-slider-thumb-focus-outline, 0.143rem dashed #000);
+    outline-offset: 0.071rem;
```
There should be a CSS variable for this value.

nikhitasharma replied:
Added

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+  .input-range:disabled::-ms-thumb {
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background: var(--terra-slider-thumb-disabled-background, #007cc3);
+    cursor: default;
```
I would expect the cursor to be on the input component not the thumb.  The disabled selector should also be on the input component instead of the thumb.  You can click on the track and cause behaviors to occur so the current design does not seem accurate in that aspect.

nikhitasharma replied:
Updated cursor to be on input component.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+
+  .input-range {
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
+    background-color: transparent;
+    margin-left: var(--terra-slider-input-range-margin-left, 0.5rem);
+    margin-right: var(--terra-slider-input-range-margin-right, 0.5rem);
+    min-height: 3.143rem;
+    min-width: 17.143rem;
+    width: 70%;
+
+    &:disabled {
+      opacity: 30%;
+    }
```
All of these values should have CSS variables.  We have more themes than called out in this component so we cannot base the logic off that.  We must account for the Cerner clinical theme and upcoming Redwood theme.

nikhitasharma replied:
Added

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+    height: var(--terra-slider-track-height, 0.43rem);
+  }
+
+  .input-range::-moz-range-thumb {
```
Aren't you able to combine some of these selectors and their corresponding styles so that there is not so much duplication?

nikhitasharma replied:
Logged UXPLATFORM-9778 for future css optimizations.

cm9361 commented:
There is still a blue border around the slider when using tab to focus the control.

cm9361 commented on packages/terra-slider/tests/wdio/__snapshots__/reference/terra-default-theme/en/chrome_large/slider-spec/keyboard-focus.png:
This screen shot does not seem to match the UX callouts.

nikhitasharma replied:
Updated to remove focus outline.

cm9361 commented:
> > You have a blue border around your slider when it has focus. The UX callouts do not specify this behavior. It seems like the border should not exists and the outline just around the slider thumb.
> 
> There is a consistent border color for all states, including focus. I have removed styling wherever overriding is unnecessary and if background color and border colors are the same.

I think it might be consistent but also incorrect.  You can update the CSS styles so that this border does not exists.  It does not seem to follow our design patterns.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+    white-space: initial;
+  }
+
+  .slider-min-label {
+    margin-right: var(--terra-slider-min-label-margin-right, 0.5rem);
+  }
+
+  .slider-max-label {
+    margin-left: var(--terra-slider-max-label-margin-left, 0.5rem);
+  }
```
Instead of creating these two classes and styles, why not just add the margin or padding to the slider element to the left and right?  That way you only have to apply the style to one element.

nikhitasharma replied:
Discussed offline, and concluded that its best to keep it this way.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+        <span className={cx('label')} aria-hidden="true">
+          {label}
+        </span>
+        <span className={cx('slider-label', 'slider-min-label')} aria-hidden="true">
+          {minLabel}
+        </span>
+        <input className={cx('input-range')} type="range" ref={this.sliderRef} aria-label={label} aria-describedby={descriptionId} value={this.state.value} disabled={isDisabled} min={minimumValue} max={maximumValue} onChange={this.handleOnChange} />
+        <span className={cx('slider-label', 'slider-max-label')} aria-hidden="true">
+          {maxLabel}
+        </span>
```
Wouldn't this be a good place to use the Terra Arrange component?  Could it help simplify your styles?

nikhitasharma replied:
Logged UXPLATFORM-9778 for future css optimizations.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  handleOnChange(event) {
+    this.setState({ value: event.currentTarget.value });
```
If the value has already been changed in the component, why do we need to set state?

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+        <span className={cx('slider-label', 'slider-min-label')} aria-hidden="true">
+          {minLabel}
+        </span>
+        <input className={cx('input-range')} type="range" ref={this.sliderRef} aria-label={label} aria-describedby={descriptionId} value={this.state.value} disabled={isDisabled} min={minimumValue} max={maximumValue} onChange={this.handleOnChange} />
```
The value should not be a state variable.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+    margin-left: var(--terra-slider-input-range-margin-left, 0.5rem);
+    margin-right: var(--terra-slider-input-range-margin-right, 0.5rem);
+    min-height: 3.143rem;
+    min-width: 17.143rem;
```
Do we really want the minimum width to be so large?  Are we considering mobile devices?  It seems like the consumer should be able to control the width by the size of the parent container.

nikhitasharma replied:
Requested @eawww  to double check on this. But I am guessing this is what they need on callouts.

nikhitasharma replied:
Looks like the concern was around the width being too small and the thumb movement being minimal those cases. That's why an optimal min-width is suggested here.

cm9361 commented on packages/terra-slider/package.json:
```diff
@@ -0,0 +1,67 @@
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "keycode-js": "^3.1.0",
```
It does not look like this dependency is being used.

nikhitasharma replied:
removed

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+        <span className={cx('slider-label', 'slider-max-label')} aria-hidden="true">
+          {maxLabel}
+        </span>
+        <VisuallyHiddenText id={descriptionId} text={VisuallyHiddenTextValue} />
```
Is this element actually required if we use an aria label?

eawww reviewed (approved):
Perfecto!

nikhitasharma commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,131 @@
+        <span className={cx('slider-label', 'slider-min-label')} aria-hidden="true">
+          {minLabel}
+        </span>
+        <input className={cx('input-range')} type="range" ref={this.sliderRef} aria-label={label} aria-describedby={descriptionId} value={this.state.value} disabled={isDisabled} min={minimumValue} max={maximumValue} onChange={this.handleOnChange} />
```
aria-describedby seems to be required based on UX inputs.

nikhitasharma commented:
> > > You have a blue border around your slider when it has focus. The UX callouts do not specify this behavior. It seems like the border should not exists and the outline just around the slider thumb.
> > 
> > 
> > There is a consistent border color for all states, including focus. I have removed styling wherever overriding is unnecessary and if background color and border colors are the same.
> 
> I think it might be consistent but also incorrect. You can update the CSS styles so that this border does not exists. It does not seem to follow our design patterns.

You are right. I initially thought you were talking about the thumb border but I now understand. I have updated this.

chrismichalewicz commented:
Looks good, great work! @nikhitasharma

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,122 @@
+
+const defaultProps = {
+  isDisabled: false,
+  onChange: undefined,
```
You should not need this as a default value.  If the prop is not passed, the onChange function will be undefined.

nikhitasharma replied:
Done

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,122 @@
+  * @private
+  * The intl object containing translations. This is retrieved from the context automatically by injectIntl.
+  */
+  intl: PropTypes.shape({ formatMessage: PropTypes.func, locale: PropTypes.string }),
```
I think we usually use this format for intl
intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,

nikhitasharma replied:
Updated

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,122 @@
+    })
+    : intl.formatMessage({ id: 'Terra.slider.ariaDescribedByTextWithoutLabels' }, { minimumValue, maximumValue });
+  const descriptionId = uuidv4();
+  const [value, setValue] = useState(props.value);
```
This should not be a state variable.  We want to consumers to be able to change the value as necessary.

nikhitasharma replied:
Discussed offline and concluded that the state is required to manage sliders internal updates, but a useEffect was added to update state when props change externally.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,122 @@
+    maximumValue,
+    label,
+    minimumLabel,
+    maximumLabel,
```
onChange should be part of this deconstruction.

nikhitasharma replied:
Added

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,122 @@
+  const handleOnChange = (event) => {
+    setValue(event.currentTarget.value);
+    if (props.onChange) {
+      props.onChange(event);
```
I wouldn't expect us to return the entire event to our consumers.  We should be returning specific information.  In this case, we probably want to return the value.  This could be an object that consumers use to obtain the value and update their state.

nikhitasharma replied:
I see the norm of passing the entire event in multiple terra components. But I changed this to make it more efficient.

cm9361 commented on packages/terra-slider/src/Slider.jsx:
```diff
@@ -0,0 +1,122 @@
+  const sliderRef = useRef();
+  const minLabel = minimumLabel || minimumValue;
+  const maxLabel = maximumLabel || maximumValue;
+  const visuallyHiddenTextValue = (minimumLabel && maximumLabel)
+    ? intl.formatMessage({ id: 'Terra.slider.ariaDescribedByTextWithLabels' }, {
+      minimumLabel, maximumLabel, minimumValue, maximumValue,
+    })
+    : intl.formatMessage({ id: 'Terra.slider.ariaDescribedByTextWithoutLabels' }, { minimumValue, maximumValue });
```
Wouldn't the calculation of this message need to be in a useEffect?  If consumers update the value, minimum value, maximum value, etc, I would think the screen reader should announce that.

nikhitasharma replied:
No, because the values used in calculation would remain the same when user interacts with slider.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+      opacity: var(--terra-slider-input-range-disabled-opacity, 30%);
+    }
+    
+    &:focus {
```
Wasn't this actually an issue for the :focus-visible state?

nikhitasharma replied:
This works as well.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  .input-range {
+    -webkit-appearance: none; /* stylelint-disable-line property-no-vendor-prefix, plugin/no-unsupported-browser-features */
```
I see us disabling linter warnings.  Why do we need to change the webkit appearance for the input?

nikhitasharma replied:
To remove the default webkit styling and add in our custom styling.

cm9361 commented on packages/terra-slider/src/Slider.module.scss:
```diff
@@ -0,0 +1,203 @@
+    border-width: var(--terra-slider-thumb-border-width, 0.071rem);
+    box-shadow: var(--terra-slider-thumb-box-shadow, 0 0 0.286rem 0 rgba(0, 0, 0, 0.25));
+    height: var(--terra-slider-thumb-height, 1.715rem);
+    margin-top: -0.57rem;
```
Looks like you missed the theme variable here.

cm9361 commented:
We should have an example and/or test that covers the consumer change the value of the Slider component via its value prop.

nikhitasharma commented:
> We should have an example and/or test that covers the consumer change the value of the Slider component via its value prop.

Added.
